runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/ActiveWorkState.java

```diff
@@ -108,9 +132,19 @@ synchronized ActivateWorkResult activateWorkForKey(ShardedKey shardedKey, Work w
     }
 
     // Ensure we don't already have this work token queued.
-    for (Work queuedWork : workQueue) {
+    Iterator<Work> workQueueIterator = workQueue.iterator();
+    while (workQueueIterator.hasNext()) {
+      Work queuedWork = workQueueIterator.next();
       if (queuedWork.getWorkItem().getWorkToken() == work.getWorkItem().getWorkToken()) {
-        return ActivateWorkResult.DUPLICATE;
+        // Work is a duplicate.
+        if (queuedWork.getWorkItem().getCacheToken() == work.getWorkItem().getCacheToken()) {
```

Comment:
could be cleaner to use WorkDedupKey or whatever it is renamed and call methods like  equals or maybe something like knownEarlierThan

if the cache token is the same, we'd prefer to keep the later work item # since they are increasing from windmill within a worker. Otherwise we can just take the last observed.

Reply:
should we add the startTime to the WorkDedupeKey/WorkId then?

Reply:
realized we don't need to do that since `Work` has the `startTime`